pkg/controller/pubsub/pubsub_init.go

```diff
@@ -0,0 +1,134 @@
+package pubsub
+
+import (
+	"context"
+	"fmt"
+
+	constraintclient "github.com/open-policy-agent/frameworks/constraint/pkg/client"
+	"github.com/open-policy-agent/frameworks/constraint/pkg/externaldata"
+	"github.com/open-policy-agent/gatekeeper/pkg/expansion"
+	"github.com/open-policy-agent/gatekeeper/pkg/logging"
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation"
+	"github.com/open-policy-agent/gatekeeper/pkg/pubsub"
+	psClient "github.com/open-policy-agent/gatekeeper/pkg/pubsub/client"
+	"github.com/open-policy-agent/gatekeeper/pkg/readiness"
+	"github.com/open-policy-agent/gatekeeper/pkg/watch"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+var log = logf.Log.WithName("controller").WithValues(logging.Process, "pubsub_controller")
+
+type Adder struct {
+	PubsubSystem *pubsub.System
+}
+
+func (a *Adder) Add(mgr manager.Manager) error {
+	if len(psClient.Providers()) == 0 {
+		return nil
+	}
+	r := newReconciler(mgr, a.PubsubSystem)
+	return add(mgr, r)
+}
+
+func (a *Adder) InjectOpa(_ *constraintclient.Client) {}
+
+func (a *Adder) InjectWatchManager(_ *watch.Manager) {}
+
+func (a *Adder) InjectControllerSwitch(_ *watch.ControllerSwitch) {}
+
+func (a *Adder) InjectTracker(_ *readiness.Tracker) {}
+
+func (a *Adder) InjectMutationSystem(_ *mutation.System) {}
+
+func (a *Adder) InjectExpansionSystem(_ *expansion.System) {}
+
+func (a *Adder) InjectProviderCache(_ *externaldata.ProviderCache) {}
+
+func (a *Adder) InjectPubsubSystem(pubsubSystem *pubsub.System) {
+	a.PubsubSystem = pubsubSystem
+}
+
+type Reconciler struct {
+	client.Client
+	scheme *runtime.Scheme
+	system *pubsub.System
+}
+
+func newReconciler(mgr manager.Manager, system *pubsub.System) *Reconciler {
+	return &Reconciler{
+		Client: mgr.GetClient(),
+		scheme: mgr.GetScheme(),
+		system: system,
+	}
+}
+
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	c, err := controller.New("pusbub-config-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	err = c.Watch(
+		&source.Kind{Type: &corev1.Secret{}},
+		&handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return c.Watch(
+		&source.Kind{Type: &corev1.ConfigMap{}},
```

Comment:
If we are going to be watching config maps, this controller needs to be toggle-able and default to off during alpha.

There is a real risk that watching config maps will drastically increase the memory footprint of Gatekeeper on some clusters, breaking users.

Reply:
do you mean something like [this](https://github.com/open-policy-agent/gatekeeper/pull/2538/files/6483cc45f3557236697130d1c3694314fa9463be#diff-52a2efd173e473faf47105460fa796cd07f13e991b9908e257ecd0f9fb780e0bR35)?  where if no pubsub is enabled then don't enable the controller and vice versa.

Reply:
Yeah, that's the right kind of code pattern, though I'm thinking a single flag to enable/disable the system as a whole is a better fit, pending the outcome of this thread:

https://github.com/open-policy-agent/gatekeeper/pull/2538#discussion_r1145554037

which is asking whether there is any use for a flag that selectively enables providers.

Reply:
So, instead of a flag to enable specific provider, a flag to enable/disable pubsub and user config to indicate which provider to use for pubsub. Is this interpretation correct?, if so I think that makes sense.

Reply:
Introduced `--enable-pub-sub` flag to enable/disable controller watching for configs.